codalab/worker/worker.py

```diff
@@ -238,10 +218,45 @@ def cleanup(self):
         logger.info("Stopped Worker. Exiting")
 
     def signal(self):
-        # When the terminate flag is False, set the stop flag to stop running
-        # the worker without changing the status of existing running bundles.
-        if not self.terminate:
+        # When the pass_down_termination flag is False, set the stop flag to stop running
+        # the worker without changing the status of existing running bundles. Otherwise,
+        # restage all bundles that are not in the terminal states [FINISHED, STAGED].
+        if not self.pass_down_termination:
             self.stop = True
+        else:
+            self.terminate_and_restage = True
+
+    def process_terminate_signal(self):
+        # If received pass_down_termination signal from CLI to terminate the worker,
+        # wait until all the existing unfinished bundles are restaged, then stop the worker.
+        if self.terminate_and_restage:
+            if self.restage_bundles() == 0:
+                self.stop = True
+
+    def restage_bundles(self):
+        """
+        Restage bundles not in the final states [FINISHED and STAGED] from worker to server.
+        :return: the number of restaged bundles
+        """
+        restaged_bundles = []
+        terminal_stages = [RunStage.FINISHED, RunStage.STAGED]
+        for uuid in self.runs:
+            run_state = self.runs[uuid]
+            if run_state.stage not in terminal_stages:
+                self.restage(uuid)
```

Comment:
Why do we have to restage `STAGED` bundles?

Reply:
hmm, not sure what you mean here... I think we try to avoid restaging the `STAGED` bundles here.